tests/test_component.py

```diff
@@ -1429,260 +1430,555 @@ class Root(Component):
 
-            def get_template_data(self, args, kwargs, slots, context):
-                return {
-                    "args": args,
-                    "kwargs": kwargs,
-                }
+            def on_render(self, context: Context, template: Optional[Template]):
+                calls.append("outer__on_render_pre")
+                html, error = yield template.render(context)  # type: ignore[union-attr]
+
+                calls.append("outer__on_render_post")
 
             # Check that modifying the context or template does nothing
-            def on_render_after(self, context: Context, template: Template, content: str) -> None:
-                # Insert value into the Context
-                context["from_on_after"] = ":)"
+            def on_render_after(
+                self,
+                context: Context,
+                template: Optional[Template],
+                html: Optional[str],
+                error: Optional[Exception],
+            ) -> None:
+                calls.append("outer__on_render_after")
 
-                # Insert text into the Template
-                template.nodelist.append(TextNode("\n---\nFROM_ON_AFTER"))
+        return Outer
 
-                nonlocal captured_content
-                captured_content = content
+    def _gen_broken_component(self):
+        class BrokenComponent(Component):
+            def on_render(self, context: Context, template: Template):
+                raise ValueError("BROKEN")
 
-        rendered = SimpleComponent.render()
+        return BrokenComponent
+
+    def test_order(self):
+        calls: List[str] = []
+
+        registry.register("slotted", self._gen_slotted_component(calls))
+        registry.register("inner", self._gen_inner_component(calls))
+        registry.register("middle", self._gen_middle_component(calls))
+        Outer = self._gen_outer_component(calls)
+
+        result = Outer.render()
 
         assertHTMLEqual(
-            captured_content,
-            """
-            args: []
-            kwargs: {}
-            ---
-            from_on_after:
-            ---
-            Hello from nested
-            <div data-djc-id-ca1bc3e data-djc-id-ca1bc40>
-                Hello from simple
-            </div>
+            result,
+            """
+            Outer start
+                Middle start
+                    Inner start
+                        Hello from slotted
+                    Inner end
+                    Middle text
+                    Inner start
+                    Inner end
+                Middle end
+                Outer text
+                Middle start
+                    Inner start
+                        Hello from slotted
+                    Inner end
+                    Middle text
+                    Inner start
+                    Inner end
+                Middle end
+            Outer end
             """,
         )
+
+        assert calls == [
```

Comment:
Here in this test we check that the order of `on_render_before()`, `on_render()` and `on_render_after()` is correct even with multiple components.